File: src/com/guerbai/concurrency/parallel_process_example.py
import time
import threading

class SumTask(object):

	def __init__(self):
		self._sub_sum = 0

	def run(self, i, num):
		start = i * num
		end = start + num
		sub_sum = 0
		while start < end:
			sub_sum += start
			start += 1
		self._sub_sum = sub_sum
	
	def get_result(self):
		return self._sub_sum


def measure_time_cost(thread_nums):
	nums = 99999999
	num_per_thread = int((nums + 1) / thread_nums)
	thread_list = [None] * thread_nums
	task_list = [None] * thread_nums
	start_at = time.time()
	for i in range(thread_nums):
		ct = SumTask()
		thread_list[i] = threading.Thread(target=ct.run, args=(i, num_per_thread))
		thread_list[i].start()
		task_list[i] = ct
	for i in range(thread_nums):
		thread_list[i].join()
	end_at = time.time()
	result = 0
	for i in range(thread_nums):
		result += task_list[i].get_result()
	print (result)
	return end_at - start_at


def main():
	thread_num1 = 1
	thread_num2 = 2
	thread_num4 = 4
	thread_num8 = 8
	print ("sum_task with thread_num1 cost time: " + str(measure_time_cost(thread_num1)) + 's in Python version.')
	print ("sum_task with thread_num2 cost time: " + str(measure_time_cost(thread_num2)) + 's in Python version.')
	print ("sum_task with thread_num4 cost time: " + str(measure_time_cost(thread_num4)) + 's in Python version.')
	print ("sum_task with thread_num4 cost time: " + str(measure_time_cost(thread_num4)) + 's in Python version.')
	

if __name__ == '__main__':
	main()

File: src/com/guerbai/concurrency/ParallelProcessExample.java
package com.guerbai.concurrency;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.*;

import static java.lang.System.currentTimeMillis;

class SumTask implements Callable<Long> {

    private int i;
    private long num;

    public SumTask(int i, long num) {
        this.i = i;
        this.num = num;
    }

    @Override
    public Long call() throws Exception {
        long start = i * num;
        long end = start + num;
        long sum = 0;
        for (long i=start; i < end; i++) {
            sum += i;
        }
        return sum;
    }

}

public class ParallelProcessExample {

    private static long measureTimeCost(int threadNum) {
        int nums = Integer.MAX_VALUE;
        ExecutorService executor = Executors.newFixedThreadPool(threadNum);
        List<Future<Long>> sumTasks = new ArrayList<>();
        long numPerThread = ((long)nums+1) / threadNum;
        long startAt = currentTimeMillis();
        for (int i=0; i<threadNum; i++) {
            sumTasks.add(executor.submit(new SumTask(i, numPerThread)));
        }
        long result = 0;
        try {
            for (int i=0; i<threadNum; i++) {
                result += sumTasks.get(i).get();
            }
        } catch (InterruptedException | ExecutionException e) {
            e.printStackTrace();
        }
        System.out.println(result);
        long endAt = currentTimeMillis();
        executor.shutdown();
        return endAt - startAt;
    }

    public static void main(String[] args) {

        int threadNum1 = 1;
        int threadNum2 = 2;
        int threadNum4 = 4;
        int threadNum8 = 8;
        System.out.println("SumTask with threadNum1 cost time: " + (float) (measureTimeCost(threadNum1)) / 1000 + "s in Java version.");
        System.out.println("SumTask with threadNum2 cost time: " + (float) (measureTimeCost(threadNum2)) / 1000 + "s in Java version.");
        System.out.println("SumTask with threadNum4 cost time: " + (float) (measureTimeCost(threadNum4)) / 1000 + "s in Java version.");
        System.out.println("SumTask with threadNum8 cost time: " + (float) (measureTimeCost(threadNum8)) / 1000 + "s in Java version.");
    }

}
